File: JPA/studyolle/src/test/java/me/kzv/olle/account/AccountControllerTest.java
package me.kzv.olle.account;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.client.AutoConfigureWebClient;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.mail.SimpleMailMessage;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.test.web.servlet.MockMvc;

import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.then;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest // (webEnvironment = ) // @AutoConfigureWebClient 실제 포트 띄우기 이걸로 서블릿 띄워서 실제로 확인하는 방법도 있음
@AutoConfigureMockMvc
class AccountControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private AccountRepository accountRepository;

    @MockBean
    JavaMailSender mailSender;

    @DisplayName("회원 가입 화면 보이는지 테스트")
    @Test
    void signUpForm() throws Exception {
        mockMvc.perform(get("/sign-up"))
                .andDo(print())
                .andExpect(status().isOk())
                .andExpect(view().name("account/sign-up"))
                .andExpect(model().attributeExists("signUpForm"))
        ;

    }

    @DisplayName("회원 가입 처리 - 입력값 오류")
    @Test
    void signUpSubmit_with_wrong_input() throws Exception {
        mockMvc.perform(post("/sign-up")
                .param("nickname", "kkk")
                .param("email", "email@email.com")
                .param("password", "12345678")
                .with(csrf()))
//                .andExpect(status().isOk())
//                .andExpect(view().name("account/sign-up")); // 입력값 오류시
                .andExpect(status().is3xxRedirection())
                .andExpect(view().name("redirect:/"));

        assertTrue(accountRepository.existsByEmail("email@email.com"));
        then(mailSender).should().send(any(SimpleMailMessage.class)); // 아무런 타입의 인스턴스를 호출하는가 - 실제로 메일이 안보내진다면 에러난다
    }
    //  spring security 적용시 csrf토큰을 기본으로 탑재하는데 form 전송시에 hidden으로 숨어서 들어간다.
}

File: toy/studyolle/src/test/kotlin/me/kzv/study/account/AccountControllerTest.kt
package me.kzv.study.account

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.BDDMockito.then
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.mail.SimpleMailMessage
import org.springframework.mail.javamail.JavaMailSender
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultHandlers.print
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*

@AutoConfigureMockMvc
@SpringBootTest
class AccountControllerTest{
    @Autowired lateinit var mockMvc: MockMvc
    @Autowired lateinit var accountRepository: AccountRepository
    @MockBean lateinit var javaMailSender: JavaMailSender

    @Test
    fun `회원 가입 화면 보이는지 테스트`() {
        //given

        //when

        //then
        mockMvc.perform(get("/sign-up"))
            .andDo(print())
            .andExpect(status().isOk)
            .andExpect(view().name("account/sign-up"))
            .andExpect(model().attributeExists("signupForm"))
    }

    @Test
    fun `회원 가입 처리 - 입력값 오류`() {
        //given

        //when

        //then
        mockMvc.perform(
            post("/sign-up")
                .param("nickname", "yul")
                .param("email", "email@email.com")
                .param("password", "12345")
                .with(csrf())
        )
//            .andExpect(status().isOk)
            .andExpect(status().is3xxRedirection)
            .andExpect(view().name("account/sign-up"))

        assertTrue(accountRepository.existsByEmail("email@email.com"))
        then(javaMailSender).should().send(any(SimpleMailMessage::class.java))
        // javaMailSender 는 smtp 서버와 연결만 해주면 되므로 인터페이스만 관리하면 된다 - 실제로 이메일을 보내는 어플리케이션은 gmail 과 같은 smtp 서버
    }

}